presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestDistributedQueries.java

```diff
@@ -714,17 +721,42 @@ public void testQueryLoggingCount()
     {
         QueryManager queryManager = ((DistributedQueryRunner) getQueryRunner()).getCoordinator().getQueryManager();
         executeExclusively(() -> {
+            assertTrueWithTimeout(
+                    () -> queryManager.getAllQueryInfo().stream().allMatch(QueryInfo::isFinalQueryInfo),
+                    new Duration(1, MINUTES),
+                    "There are still queries in process: %s", queryManager.getAllQueryInfo());
+
             long beforeQueryCount = queryManager.getStats().getCompletedQueries().getTotalCount();
             assertUpdate("CREATE TABLE test_query_logging_count AS SELECT 1 foo_1, 2 foo_2_4", 1);
             assertQuery("SELECT foo_1, foo_2_4 FROM test_query_logging_count", "SELECT 1, 2");
             assertUpdate("DROP TABLE test_query_logging_count");
             assertQueryFails("SELECT * FROM test_query_logging_count", ".*Table .* does not exist");
-
-            long queryCount = queryManager.getStats().getCompletedQueries().getTotalCount() - beforeQueryCount;
-            assertEquals(queryCount, 4, format("expected [%s] but found [%s]", 4, queryCount));
+            // TODO: Figure out a better way of synchronization
+            assertTrueWithTimeout(
+                    () -> queryManager.getStats().getCompletedQueries().getTotalCount() - beforeQueryCount == 4,
+                    new Duration(1, MINUTES),
+                    "expected [%s] but found [%s]", 4, queryManager.getStats().getCompletedQueries().getTotalCount() - beforeQueryCount);
```

Comment:
Message args will be computed immediately, while condition (whether there are 4) will be computed repeatedly for 1 minute. If, for example 3 queries (out of 4 expected) completed within the timeout, the exception message will still say `expected 4 but found 0` instead of `expected 4 but found 3`. This is misleading, especially if one is to debug test failures.

It will be better if you change it to `Supplier< args >`. However, this leaves the chance that exception will be raised with err msg like `expected 4 got 4`. Maybe checking condition could be lambda's responsibility? Like

```
assertWithTimeout(Runnable checkCondition, Duration ...)
```

Or, alternatively, 
```
assertEqualWithTimeout(T expected, Supplier<T> actual, Duration ...)
```
?